src/service/app.py

```diff
@@ -60,37 +60,52 @@ def create_app(noop=False):
         }
     )
     app.include_router(_router)
+
     async def build_app_wrapper():
         await _build_app(app, cfg)
-
     app.add_event_handler("startup", build_app_wrapper)
+
+    async def clean_app_wrapper():
+        await _clean_app(app)
+    app.add_event_handler("shutdown", clean_app_wrapper)
     return app
 
 
 async def _build_app(app, cfg):
     app.state.kb_auth = await kb_auth.create_auth_client(cfg.auth_url, cfg.auth_full_admin_roles)
-    cli = aioarango.ArangoClient(hosts=cfg.arango_url)
-    if cfg.create_db_on_startup:
-        sysdb = await _get_arango_db(cli, "_system", cfg)
-        try:
-            await sysdb.create_database(cfg.arango_db)
-        except aioarango.exceptions.DatabaseCreateError as e:
-            if e.error_code != 1207:  # duplicate name error, ignore, db exists
-                raise
-    db = await _get_arango_db(cli, cfg.arango_db, cfg)
-    app.state.storage = await create_storage(
-        db, create_collections_on_startup=cfg.create_db_on_startup)
-    sys.stdout.flush()
+    cli, storage = await _build_storage(cfg)
+    app.state.arango_cli = cli
+    app.state.storage = storage
+
 
+async def _clean_app(app):
+    if app.state.arango_cli:
+        await app.state.arango_cli.close()
 
-_BACKOFF = [0, 1, 2, 5, 10, 30]
+
+async def _build_storage(cfg):
+    if cfg.dont_connect_to_external_services:
+        return False, False
+    cli = aioarango.ArangoClient(hosts=cfg.arango_url)
+    try:
+        if cfg.create_db_on_startup:
+            sysdb = await _get_arango_db(cli, "_system", cfg)
+            try:
+                await sysdb.create_database(cfg.arango_db)
+            except aioarango.exceptions.DatabaseCreateError as e:
+                if e.error_code != ARANGO_ERR_NAME_EXISTS:  # ignore, db exists
+                    raise
+        db = await _get_arango_db(cli, cfg.arango_db, cfg)
+        storage = await create_storage(db, create_collections_on_startup=cfg.create_db_on_startup)
+        return cli, storage
+    except:
+        await cli.close()
+        raise
```

Comment:
will it provide us with enough traceback to troubleshoot?

Reply:
I don't understand the concern - the `raise` will raise the exact same error the was caught so we're not losing any info...?

Reply:
yea. yea. you are right.